src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,786 @@
+
+    Download the HIFLD planning area geometries if they are not already available
+    locally. Store the data in "local/hifld/electric_planning_areas.gdb" under the PUDL
+    datastore's top level directory, as they are not integrated into the core PUDL data
+    management. The data is in the geodatabase format.
 
-    The function returns a stacked dataframe with mapping of each intermediate
-    geometry column with every non-intersecting source geometry column
-    quantifying the intersecting area.
+    Read the data into a goepandas.GeoDataFrame, and convert the data types as needed.
+    All columns are assigned nullable types.
 
     Args:
-        gdf_intermediate (GeoDataframe): intermediate geodataframe
-        gdf_source (GeoDataframe): usually source dataframe (i.e. demand)
-        gdf_intermediate_col (str): column to identify gdf_intermediate
-        gdf_source_col(str): column to identify for gdf_source
-        file_save (bool): save file option
+        pudl_settings (dict): A dictionary of PUDL settings, including the path to the
+            datastore.
 
     Returns:
-        pandas.DataFrame: DataFrame used to map and scale attribute from one
-        set of geometries to another.
-
-        Columns:
-            intermediate_index: unique index for intermediate geometry
-            source_index: unique index for source geometry
-            intermediate_intersection_fraction: area as fraction of
-            intermediate geometry
-            source_intersection_fraction: area as fraction of source geometry
+        geopandas.GeoDataFrame: containing the HIFLD planning area geometries and all
+        associated data columns.
 
     """
-    new_df = geopandas.overlay(
-        gdf_intermediate[[gdf_intermediate_col, geom_intermediate]],
-        gdf_source[[gdf_source_col, geom_source]],
-        how='intersection'
+    hifld_pa_url = "https://opendata.arcgis.com/datasets/7d35521e3b2c48ab8048330e14a4d2d1_0.gdb"
+    hifld_dir = pathlib.Path(pudl_settings["data_dir"]) / "local/hifld"
+    hifld_dir.mkdir(parents=True, exist_ok=True)
+    hifld_pa_zipfile = hifld_dir / "electric_planning_areas.gdb.zip"
+    hifld_pa_gdb_dir = hifld_dir / "electric_planning_areas.gdb"
+
+    if not hifld_pa_gdb_dir.is_dir():
+        logger.warning("No Planning Area GeoDB found. Downloading from HIFLD.")
+        # Download to appropriate location
+        pudl.helpers.download_zip_url(hifld_pa_url, hifld_pa_zipfile)
+        # Unzip because we can't use zipfile paths with geopandas
+        with zipfile.ZipFile(hifld_pa_zipfile, 'r') as zip_ref:
+            zip_ref.extractall(hifld_dir)
+            # Grab the UUID based directory name so we can change it:
+            extract_root = hifld_dir / \
+                pathlib.Path(zip_ref.filelist[0].filename).parent
+        extract_root.rename(hifld_pa_gdb_dir)
+    else:
+        logger.info("We've already got the planning area GeoDB.")
```

Comment:
Yeah, it's definitely a pattern, but I wasn't sure how much additional design & generalization work it was worth putting into it, since 714 will get pulled into PUDL, and I'm not sure how durable the Census / HIFLD data usage will be. But I'm sure we'll keep using external non-PUDL data, and we should develop a standard-ish interface to it.

Reply:
That last part there is what I am thinking. I assume this will be a type of thing we will have to do more than once and having a standardized way to do it seems like a good call.

Reply:
I was just trying to write up an issue for this and... I think the right answer may be that more of us need to learn how to integrate additional archives into the datastore. Really that's what I'm avoiding here. The data aren't particularly large, in the context of what Zenodo allows you to store for free (HIFLD 10MB, FERC 714 52MB, Census 264 MB), and having a single way that these external data sources are accessed would be much better than accreting another layer of cruft on here. And then those simple data packages in the datastore can be wrapped with an ETL-lite that doesn't try and pull things into the DB, but just reads the data directly and turns it into a dataframe or geodataframe or dictionary of dataframes. This seems like a good kind of staging area for future datasets of interest that may or may not be worth (or need) full integration.

Reply:
i think what you are saying here is you'd rather not spend time making it easy to not use the datastore when pulling in temp or pudl-agacent datasets. that seems fair to me I guess. for things that are very accessible like the census data this seems... potentially very unnecessary. but if it is simple I won't hate it. lol